app/src/main/kotlin/compiler/semantic_analysis/ArgumentResolver.kt

```diff
@@ -2,17 +2,182 @@ package compiler.semantic_analysis
 
 import compiler.ast.Expression
 import compiler.ast.Function
+import compiler.ast.NamedNode
 import compiler.ast.Program
+import compiler.ast.Statement
+import compiler.ast.StatementBlock
+import compiler.ast.Variable
+import compiler.common.diagnostics.Diagnostic
 import compiler.common.diagnostics.Diagnostics
+import compiler.common.reference_collections.ReferenceHashMap
 import compiler.common.reference_collections.ReferenceMap
 
 typealias ArgumentResolutionResult = ReferenceMap<Expression.FunctionCall.Argument, Function.Parameter>
 
-object ArgumentResolver {
-    fun calculateArgumentToParameterResolution(
-        ast: Program,
-        diagnostics: Diagnostics
-    ): ArgumentResolutionResult {
-        return TODO() // also add errors under Diagnostic.ArgumentResolutionError and uncomment in Compiler
+class ArgumentResolver(private val nameResolution: ReferenceMap<Any, NamedNode>, private val diagnostics: Diagnostics) {
+    private val result = ReferenceHashMap<Expression.FunctionCall.Argument, Function.Parameter>()
+    private var correctDefinitions = true
+
+    companion object {
+        fun calculateArgumentToParameterResolution(
+            program: Program,
+            nameResolution: ReferenceMap<Any, NamedNode>,
+            diagnostics: Diagnostics
+        ): ArgumentResolutionResult {
+            val resolver = ArgumentResolver(nameResolution, diagnostics)
+
+            resolver.checkFunctionDefinitions(program)
+            if (resolver.correctDefinitions)
+                resolver.checkFunctionCalls(program)
+
+            return resolver.result
+        }
+    }
+
+    private fun checkFunctionCall(functionCall: Expression.FunctionCall) {
+        var foundNamedArgument = false
+        for (argument in functionCall.arguments) {
+            if (argument.name != null) {
+                foundNamedArgument = true
+            } else if (foundNamedArgument) {
+                diagnostics.report(Diagnostic.ArgumentResolutionError.PositionalArgumentAfterNamed(functionCall))
+                return
+            }
+        }
+
+        val function: Function = nameResolution[functionCall] as Function
+        val parameterNames = function.parameters.map { it.name }
+        val numberOfParameters = parameterNames.size
+        val isMatched = parameterNames.map { false }.toMutableList()
+
+        for ((index, argument) in functionCall.arguments.withIndex()) {
+            if (argument.name == null) {
+                if (index >= numberOfParameters) {
+                    diagnostics.report(Diagnostic.ArgumentResolutionError.TooManyArguments(functionCall))
```

Comment:
It would be more clear if we checked before the loop if `arguments.size > numberOfParameters`.